File: Benchmark/Cpp/MatrixMultiplier.cpp
#include <vector>
#include <iostream>
#include <chrono>
#include <cstdlib>
#include <fstream>

using namespace std;


class MatrixMult {

public:

    vector<vector<int>> MatrixMultiply(
        vector<vector<int>>& A,
        vector<vector<int>>& B
    ) {

        int n = A.size();

        vector<vector<int>> C(n, vector<int>(n, 0));

        for (int row = 0; row < n; row++) {
            for (int col = 0; col < n; col++) {
                for (int k = 0; k < n; k++) {
                    C[row][col] += A[row][k] * B[k][col];
                }
            }
        }

        return C;
    }
};


vector<vector<int>> generateMatrix(int n) {

    vector<vector<int>> matrix(n, vector<int>(n));

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            matrix[i][j] = rand() % 10;
        }
    }

    return matrix;
}


int main() {

    fstream output("Cpp_results.txt");

    MatrixMult multiplier;

    int matrix_sizes[] = {16, 32, 64, 128, 256};

    int num_runs = 5;

    output << "C++ Matrix Multiplication Benchmark Results:" << endl;

    for (int n : matrix_sizes) {

        double total_time = 0;

        vector<vector<int>> A = generateMatrix(n);
        vector<vector<int>> B = generateMatrix(n);

        cout << "\n" << n << "x" << n << endl;
        output << n << "x" << n << endl;

        for (int run = 0; run < num_runs; run++) {

            auto start = chrono::high_resolution_clock::now();

            vector<vector<int>> result =
                multiplier.MatrixMultiply(A, B);

            auto end = chrono::high_resolution_clock::now();

            chrono::duration<double, milli> elapsed =
                end - start;

            total_time += elapsed.count();

            cout << "Run " << run + 1
                 << ": " << elapsed.count()
                 << " ms" << endl;


            output << "Run " << run + 1
                << ": " << elapsed.count()
                << " ms" << endl;
        }

        double average_time = total_time / num_runs;

        cout << "Average: "
             << average_time
             << " ms" << endl;

        output << n << "x" << n
               << " Average: "
               << average_time
               << " ms" << endl;
        output << endl;
    }

    output.close();

}

File: Benchmark/Python/MatrixMultiplier.py
import random
import time


class MatrixMult:

    def MatrixMultiply(self, A, B):
        n = len(A)

        C = [[0 for _ in range(n)] for _ in range(n)]

        for row in range(n):
            for col in range(n):
                for k in range(n):
                    C[row][col] += A[row][k] * B[k][col]

        return C


def generateMatrix(n):
    matrix = []

    for i in range(n):
        row = []

        for j in range(n):
            row.append(random.randint(0, 9))

        matrix.append(row)

    return matrix


def main():

    output = open("py_results.txt", "w")

    multiplier = MatrixMult()

    matrix_sizes = [16, 32, 64, 128, 256]

    num_runs = 5

    output.write("Python Matrix Multiplication Benchmark Results:\n\n")

    for n in matrix_sizes:

        total_time = 0

        A = generateMatrix(n)
        B = generateMatrix(n)

        print(f"\n{n}x{n}")
        output.write(f"{n}x{n}\n")

        for run in range(num_runs):

            start = time.perf_counter()

            result = multiplier.MatrixMultiply(A, B)

            end = time.perf_counter()

            elapsed = (end - start) * 1000

            total_time += elapsed

            # Terminal
            print(f"Run {run + 1}: {elapsed:.6f} ms")

            # File
            output.write(
                f"Run {run + 1}: {elapsed:.6f} ms\n"
            )

        # Calculate average
        average_time = total_time / num_runs

        # Terminal
        print(f"Average: {average_time:.6f} ms")

        # File
        output.write(
            f"Average: {average_time:.6f} ms\n"
        )

        output.write("\n")

    output.close()


if __name__ == "__main__":
    main()